### PR TITLE
Add script for updating wpt interfaces

### PR DESCRIPTION
**Pull request by lukebjerring**
@mdittmer for Journal reading.
@foolip for curation of spec list.

Example journal:
https://gist.github.com/lukebjerring/b23bc4006230fd0a1c5991037ba491c2

### CODE REVIEW

**Comment by lukebjerring** [comment]
@foolip 

    The following 22 specs were not resolved:
    https://w3c.github.io/pointerevents/extension.html
    https://w3c.github.io/gamepad/extensions.html
    https://wicg.github.io/shape-detection-api/text.html
    https://w3c.github.io/webcomponents/spec/custom/
    https://w3c.github.io/webcomponents/spec/imports/
    https://w3c.github.io/webcomponents/spec/shadow/
    https://immersive-web.github.io/webxr/spec/latest/
    https://svgwg.org/svg2-draft/coords.html
    https://www.khronos.org/registry/webgl/specs/latest/1.0/
    https://www.khronos.org/registry/webgl/specs/latest/2.0/
    https://svgwg.org/svg2-draft/styling.html
    https://svgwg.org/svg2-draft/interact.html
    https://svgwg.org/svg2-draft/types.html
    https://svgwg.org/svg2-draft/struct.html
    https://svgwg.org/svg2-draft/shapes.html
    https://svgwg.org/svg2-draft/embedded.html
    https://svgwg.org/svg2-draft/painting.html
    https://svgwg.org/svg2-draft/paths.html
    https://svgwg.org/svg2-draft/linking.html
    https://svgwg.org/svg2-draft/text.html
    https://svgwg.org/svg2-draft/pservers.html
    https://svgwg.org/specs/animations/

File: main/update-wpt.js [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+// Script hacked together to load raw data for exploration in a REPL.
+
+require('foam2');
+const path = require('path');
+const fs = require('fs');
+const URL = require('url').URL
+const flags = require('flags');
+
+require(path.resolve(__dirname, '..', 'data/raw.js'));
```
Comment (mdittmer) [inline main/update-wpt.js]: Where does this come from?

File: main/update-wpt.js [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+require(path.resolve(__dirname, 'specs.js'));
```
Comment (mdittmer) [inline main/update-wpt.js]: Could we just make it `data/specs.json` and use `fs.readFileSync(... JSON.parse(...))`?

File: main/update-wpt.js [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+flags.defineString('journal', './data/IDLFragmentExtractorRunner-WebSpec-journal.js', 'File containing the webidl-diff journal output.');
```
Comment (mdittmer) [inline main/update-wpt.js]: Didn't know about this library *the more you know*. I've always used `yargs`.

File: main/update-wpt.js [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    && !result.pathname.endsWith('/')) {
+    result.pathname = result.pathname + '/';
+  }
+  return result;
+}
+
+function sameSpec(spec, href) {
+  let n = normalizedURL(href).href;
+  return normalizedURL(spec.href).href == n
+    || spec.alt && spec.alt.find(url => normalizedURL(url).href == n);
```
Comment (mdittmer) [inline main/update-wpt.js:47]: `a.find()` truthiness check slightly dangerous (could find a falsey value), but seems unlikely.

File: main/update-wpt.js [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+async function main() {
+  if (!fs.existsSync(JOURNAL_FILE)) {
+    throw new Error(`journal file ${JOURNAL_FILE} not found`);
+  }
+  const journal = foam.dao.JDAO.create({
+    of: foam.core.FObject,
+    delegate: foam.dao.ArrayDAO.create(),
```
Comment (mdittmer) [inline main/update-wpt.js:56]: nit: Set `ArrayDAO.of` in kind.

File: main/update-wpt.js [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+  return normalizedURL(spec.href).href == n
+    || spec.alt && spec.alt.find(url => normalizedURL(url).href == n);
+}
+
+async function main() {
+  if (!fs.existsSync(JOURNAL_FILE)) {
+    throw new Error(`journal file ${JOURNAL_FILE} not found`);
+  }
+  const journal = foam.dao.JDAO.create({
+    of: foam.core.FObject,
```
Comment (mdittmer) [inline main/update-wpt.js:55]: nit: This should be whatever class is in the journal.

File: main/update-wpt.js [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+      fd: require('fs').openSync(JOURNAL_FILE, 'r'),
+    }),
+  });
+  let extractedSpecs = (await journal.select()).array;
+
+  let extractedSpecsByURL = new Map();
+  extractedSpecs.forEach(i => { extractedSpecsByURL.set(normalizedURL(i.url).href, []) });
+  extractedSpecs.forEach(i => { extractedSpecsByURL.get(normalizedURL(i.url).href).push(i.contents) });
+
+  let missing = []
```
Comment (mdittmer) [inline main/update-wpt.js:75]: nit: semicolon

File: main/update-wpt.js [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+  const journal = foam.dao.JDAO.create({
+    of: foam.core.FObject,
+    delegate: foam.dao.ArrayDAO.create(),
+    journal: foam.dao.NodeFileJournal.create({
+      fd: require('fs').openSync(JOURNAL_FILE, 'r'),
+    }),
+  });
+  let extractedSpecs = (await journal.select()).array;
+
+  let extractedSpecsByURL = new Map();
```
Comment (mdittmer) [inline main/update-wpt.js]: Comment here? I'm having trouble following the two `forEach`s.

File: main/update-wpt.js [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+    let concat = snippets.join('\n\n')
+      .replace(/ +(\n|$)/g, '\n') // Drop trailing whitespace (per line)
+      .replace(/\n\n+/g, '\n\n') // More that 2 newlines => 2 newlines
+      .replace(/\n*$/, '\n');  // Finish with exactly one newline
+
+
+    let outdir = OUT_DIR ||
+      path.resolve(__dirname, '..', '..', 'web-platform-tests/interfaces');
+    if (!fs.existsSync(outdir)) {
```
Comment (mdittmer) [inline main/update-wpt.js]: nit: Maybe do this check up front.

File: main/update-wpt.js (hunk 6 above)
Comment (mdittmer) [reply main/update-wpt.js:55]: Looks like it's `org.chromium.webidl.IDLSpecFile`.

File: main/specs.js [hunk 10]
```diff
@@ -0,0 +1,1622 @@
+/**
+* A list of spec sources around the web.
```
Comment (foolip) [inline main/specs.js:2]: Can you fetch this from https://foolip.github.io/day-to-day/specs.json instead?

**Comment by foolip** [comment]
@lukebjerring, with an updated specs.json, do you still have unmatched URLs? All Chromium changes except https://chromium-review.googlesource.com/c/chromium/src/+/952670 have landed, but don't know if IDL scraper has been run since.

**Comment by mdittmer** [comment]
@lukebjerring is this abandoned? Looks like some tests need updates and there are a few comments that remain unaddressed.

**Comment by lukebjerring** [comment]
Script is made redundant if we'll be using reffy instead.

**Pull request closed without merging**